control-plane/api/v1alpha1/serviceresolver_types.go

```diff
@@ -513,17 +653,80 @@ func (in *ServiceResolver) validateEnterprise(consulMeta common.ConsulMeta) fiel
 }
 
 func (in *ServiceResolverFailover) isEmpty() bool {
-	return in.Service == "" && in.ServiceSubset == "" && in.Namespace == "" && len(in.Datacenters) == 0 && len(in.Targets) == 0 && in.Policy == nil
+	return in.Service == "" && in.ServiceSubset == "" && in.Namespace == "" && len(in.Datacenters) == 0 && len(in.Targets) == 0 && in.Policy == nil && in.SamenessGroup == ""
 }
 
-func (in *ServiceResolverFailover) validate(path *field.Path) *field.Error {
+func (in *ServiceResolverFailover) validate(path *field.Path, subset string, consulMeta common.ConsulMeta) field.ErrorList {
+	var errs field.ErrorList
 	if in.isEmpty() {
 		// NOTE: We're passing "{}" here as our value because we know that the
 		// error is we have an empty object.
-		return field.Invalid(path, "{}",
-			"service, serviceSubset, namespace, datacenters, policy, and targets cannot all be empty at once")
+		errs = append(errs, field.Invalid(path, "{}",
+			"service, serviceSubset, namespace, datacenters, policy, and targets cannot all be empty at once"))
 	}
-	return nil
+
+	if consulMeta.Partition != "default" && len(in.Datacenters) != 0 {
+		errs = append(errs, field.Invalid(path.Child("datacenters"), in.Datacenters,
+			"cross-datacenter failover is only supported in the default partition"))
+	}
+
+	errs = append(errs, in.Policy.validate(path.Child("policy"))...)
+
+	asJSON, _ := json.Marshal(in)
+	if in.SamenessGroup != "" {
+		switch {
+		case len(in.Datacenters) > 0:
+			errs = append(errs, field.Invalid(path, string(asJSON),
+				"samenessGroup cannot be set with datacenters"))
+		case in.ServiceSubset != "":
+			errs = append(errs, field.Invalid(path, string(asJSON),
+				"samenessGroup cannot be set with serviceSubset"))
+		case len(in.Targets) > 0:
+			errs = append(errs, field.Invalid(path, string(asJSON),
+				"samenessGroup cannot be set with targets"))
+		}
+	}
+
+	if len(in.Datacenters) != 0 && len(in.Targets) != 0 {
+		errs = append(errs, field.Invalid(path, string(asJSON),
+			"targets cannot be set with datacenters"))
+	}
+
+	if in.ServiceSubset != "" && len(in.Targets) != 0 {
+		errs = append(errs, field.Invalid(path, string(asJSON),
+			"targets cannot be set with serviceSubset"))
+	}
+
+	if in.Service != "" && len(in.Targets) != 0 {
+		errs = append(errs, field.Invalid(path, string(asJSON),
+			"targets cannot be set with service"))
+	}
+
+	for i, target := range in.Targets {
+		asJSON, _ := json.Marshal(target)
+		switch {
+		case target.Peer != "" && target.ServiceSubset != "":
+			errs = append(errs, field.Invalid(path.Child("targets").Index(i), string(asJSON),
```

Comment:
same comment about the json object

Reply:
resolving this convo: removed validation on subsets